### PR TITLE
[Bounties] Game Theory Module Revamp

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Review by ryanchou-dev — APPROVED** [review]
@dongliu0426 Can you also review the explanations?

Everything LGTM!

**Comment by SansPapyrus683** [comment]
will sansify later when i have the time lol

File: content/6_Advanced/Game_Theory.mdx [hunk 1]
```diff
@@ -40,10 +38,285 @@ frequency: 0
...
+<Info title="NIM game with increases">
+In some cases there could be allowed to also **add** stones to a chosen pile, so when a turn comes, you can remove or add stones. It turns out that adding doesn't change the how winning and losing states are determined i.e. adding operation doesn't affect the outcome of the game.
+</Info>
+
+
+### Demonstration
+
+To demonstrate this theory we have to show that from a state with xor-sum $0$ (losing), whatever the move is, we can only reach a state with non-zero xox-sum (winning), and from a winning state we can reach a losing state.
+
+We will assume that from a state with sum-xor $0$ we can also reach another state with xor-sum $0$. Let's pick an arbitrary pile with $y$ stones and we would like to remove $x$ stones ($x \in [1, y]$).
```
Comment (envyaims) [inline content/6_Advanced/Game_Theory.mdx:66]: why is one sum-xor and the other xor-sum

File: content/6_Advanced/Game_Theory.mdx [hunk 2]
```diff
@@ -40,10 +38,285 @@ frequency: 0
...
+
+- $n(0) = 0$
+- $n(1) = \text{mex}\{0\} = 1$
+- $n(2) = \text{mex}\{0, 1\} = 2$
+- $n(3) = \text{mex}\{0, 1, 2\} = 3$
+- $n(4) = \text{mex}\{1, 2, 3\} = 0$
+- $n(5) = \text{mex}\{2, 3, 0\} = 1$
+- $n(6) = \text{mex}\{3, 0, 1\} = 2$
+- $n(7) = \text{mex}\{0, 1, 2\} = 3$
+
```
Comment (envyaims) [inline content/6_Advanced/Game_Theory.mdx:148]: add a general case, like
n(i) = mex(n(i-1), n(i-2), (i-3))

**Review by envyaims — APPROVED** [review]
wow i love this a lot. very great work!

File: content/6_Advanced/Game_Theory.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:66]: Fixed

File: content/6_Advanced/Game_Theory.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:149]: Done

**Comment by SansPapyrus683** [comment]
i'll get around to this someday

**Comment by brebenelmihnea** [comment]
you rewritten the whole thing :))

**Comment by SansPapyrus683** [comment]
yeah

File: content/6_Advanced/Game_Theory.mdx [hunk 3]
```diff
@@ -40,10 +38,389 @@ frequency: 0
...
+		url="https://www.youtube.com/watch?v=GRlGknQEOW8"
+		starred
+	/>
+</Resources>
+
+The [Sprague-Grundy theorem](https://en.wikipedia.org/wiki/Sprague%E2%80%93Grundy_theorem) defines "nimbers" or "grundy values" as
+
+$$
+\tag{1}
+n(x) = \text{mex}\{\text{all reachable nimber values}\}
```
Comment (SansPapyrus683) [inline content/6_Advanced/Game_Theory.mdx:233]: this is a recursive definition of what nimbers are

File: content/6_Advanced/Game_Theory.mdx [hunk 4]
```diff
@@ -40,10 +38,389 @@ frequency: 0
...
+- $n(4) = \text{mex}\{1, 2, 3\} = 0$
+- $n(5) = \text{mex}\{2, 3, 0\} = 1$
+- $n(6) = \text{mex}\{3, 0, 1\} = 2$
+- $n(7) = \text{mex}\{0, 1, 2\} = 3$
+
+A nimber values equal to $0$ represents a losing state of the game; see $n(4)$ and $n(7)$.
+
+<Spoiler title="Demonstration">
+- Recall that a winning state means that in one move you can reach push the other player in a losing state. For example, $n(4)=0$ because we can't reach another losing state, like $n(0)$. Similarly $n(8) = 0$, because we can't reach $n(4)$ removing 1, 2 or 3 stones in one move.
+</Spoiler>
```
Comment (SansPapyrus683) [inline content/6_Advanced/Game_Theory.mdx:257]: why's this in a spoiler

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Game_Theory.mdx (hunk 3 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:233]: It's not recursive. It's just what contains.

File: content/6_Advanced/Game_Theory.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:257]: I'll remove it

File: content/6_Advanced/Game_Theory.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Game_Theory.mdx:233]: you haven't explained what a nimber is

Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:233]: `This means that every state of the game can be assigned a nimber value.`

I don't think it has a straight definition.

Comment (SansPapyrus683) [reply content/6_Advanced/Game_Theory.mdx:233]: wdym?

Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:233]: Ok then it's a recursive def

Comment (SansPapyrus683) [reply content/6_Advanced/Game_Theory.mdx:233]: i mean do you see how this is a problem

**Comment by brebenelmihnea** [comment]
@SansPapyrus683 better?

File: content/6_Advanced/Game_Theory.mdx (hunk 3 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx:233]: `defines "nimbers" or "grundy values" of any impartial game as the unique one-pile NIM game that the state of the game is equivalent to. These values can be computed using the recursive formula:`

I added this

**Comment by brebenelmihnea** [comment]
@SansPapyrus683 merge?

File: content/6_Advanced/Game_Theory.mdx [hunk 5]
```diff
@@ -40,10 +38,389 @@ frequency: 0
...
+</CPPSection>
+</LanguageSection>
+
+#### Strange Nim
+
+<FocusProblem problem="strange-nim" />
+
+We can apply the Sprague-Grundy theorem $(2)$. Let $g(n,k)$ be the nimber value for a particular pile. To speed up the algorithm, a few observations are necessary:
+1. $g(n, k) = n/k \iff k \mid n$
+2. $g(n, k) = g(n - \lfloor n/k \rfloor - 1, k)$
```
Comment (SansPapyrus683) [inline content/6_Advanced/Game_Theory.mdx:329]: elaborate?

File: content/6_Advanced/Game_Theory.mdx [hunk 6]
```diff
@@ -40,10 +38,389 @@ frequency: 0
...
+		url="https://cp-algorithms.com/game_theory/sprague-grundy-nim.html"	/>
+	<Resource
+		source="Youtube"
+		title="Sprague-Grundy-Theorem"
+		url="https://www.youtube.com/watch?v=GRlGknQEOW8"
+		starred
+	/>
+</Resources>
+
+The [Sprague-Grundy theorem](https://en.wikipedia.org/wiki/Sprague%E2%80%93Grundy_theorem) defines "nimbers" or "grundy values" of any impartial game as the unique one-pile NIM game that the state of the game is equivalent to. These values can be computed using the recursive formula:
```
Comment (SansPapyrus683) [inline content/6_Advanced/Game_Theory.mdx:229]: why would a nimber value of 3 or 2 be any different from a nimber value of 1 if it's a single pile and you can take as many as you want

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Game_Theory.mdx (hunk 6 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx]: Ok now I think it should be a little bit clearer

File: content/6_Advanced/Game_Theory.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx]: They are patterns in those particular grundy values.  Idk how to demonstrate.

Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx]: Added suggestion for looking after patterns.

File: content/6_Advanced/Game_Theory.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Game_Theory.mdx]: i mean i thought the theorem said all games were equivalent to a one pile game of nim and that was it

Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx]: it's ok now?

Comment (SansPapyrus683) [reply content/6_Advanced/Game_Theory.mdx]: can you wait a lil, it's finals/move out week for me lol
don't have time to do an in depth review 💀

File: content/6_Advanced/Game_Theory.mdx [hunk 7]
```diff
@@ -40,10 +38,572 @@ frequency: 0
...
+#### Explanation
+
+Sometimes, we have to make a clever observation or two to be able to apply the formula in a feasible manner.
+These observations may involve periodicity or special cases, and the latter is what we'll use.
+
+Let $g(n, k)$ be the nimber value for a particular pile with $n$ stones and an associated integer of $k$.
+Notice the two following things:
+
+1. $g(n, k) = n/k \iff k \mid n$
+2. $g(n, k) = g(n - \lfloor n/k \rfloor - 1, k)$
```
Comment (SansPapyrus683) [inline content/6_Advanced/Game_Theory.mdx:545]: i have no idea why these are true

Comment (brebenelmihnea) [reply content/6_Advanced/Game_Theory.mdx]: I wrote more datails

**Review by nwatx — APPROVED** [review]

**Review by SansPapyrus683 — APPROVED** [review]

**Pull request merged**